src/Servers/Kestrel/shared/KnownHeaders.cs

```diff
@@ -1328,6 +1328,15 @@ public unsafe bool TryHPackAppend(int index, ReadOnlySpan<byte> value, bool chec
             }}
         }}
 
+        [MethodImpl(MethodImplOptions.AggressiveOptimization)]
+        public void MergeCookies()
+        {{
+            if (_headers._Cookie.Count > 1)
+            {{
+                _headers._Cookie = string.Join(""; "", _headers._Cookie.ToArray());
+            }}
+        }}
```

Comment:
I wonder if it'd be slightly faster in the more common case when the request doesn't contain any cookie header to check `_bits` (e.g. `(_bits & 0x20000L) != 0`) first before checking `_headers._Cookie.Count > 1`. Both checks should be pretty fast, but doing the `_bits` check first might also be technically more correct because `_bits` can be unset between requests while the `_headers._Cookie` contains a value.

https://github.com/dotnet/aspnetcore/blob/0200c0571a2fa06e6f9da85dc02d2d5382c6d3af/src/Servers/Kestrel/Core/src/Internal/Http/HttpRequestHeaders.cs#L51-L68

`MergeCookies()` isn't being called between requests though, so I think the logic is probably correct already but possibly a little slower than necessary in the missing header case. 

If we don't update this to check `_bits`, we can move this to [HttpRequestHeaders.cs](https://github.com/dotnet/aspnetcore/blob/0200c0571a2fa06e6f9da85dc02d2d5382c6d3af/src/Servers/Kestrel/Core/src/Internal/Http/HttpRequestHeaders.cs) since it's not taking advantage of any templating otherwise.

We should benchmark `HTTP/2` before and after with no cookies, a single cookie header and multiple cookie headers.

---

Other thoughts,

The `IHeaderDictionary.Cookie` getter checks `_bits` first before deciding whether to return `default` instead `_headers._Cookie`, so this still has me a little worried about correctness. If `_headers._Cookie` is only non-default between requests or when there is an actual value, why are we bothering with the extra check in `IHeaderDictionary.Cookie` and the other getters? It would be so much simpler to just `return _headers._Cookie;`.

Is it just because we copied `TryGetValueFast` which has to return a bool in addition to providing a `StringValues` object? The indexer calls `TryGetValueFast`, but doesn't check the return value. We might be able to speed up our getters and indexer by just returning the values directly from the fields without checking `_bits`.

I look into this and implement it in another PR if it's safe to do this.